### PR TITLE
Remove `Maybe` from `convertPlutusData` in `(De)serialization` modules

### LINKED ISSUE #1284: Remove `Maybe` from `convertPlutusData`
It will affect  `serializeData :: forall (a :: Type). ToData a => a -> Maybe CborBytes` type - we will be able to drop the `Maybe`.

It's OK to unwrap the Maybe unsafely, because we can rely on the fact that it's an enum.

### PR DESCRIPTION
**Pull request by uhbif19**
Closes #1284.

### Pre-review checklist

- [x] All code has been formatted using our config (`make format`)
- [x] Any new API features or modification of existing behavior is covered as defined in our [test plan](https://github.com/Plutonomicon/cardano-transaction-lib/blob/develop/doc/test-plan.md)
- [x] The changelog has been updated under the `## Unreleased` header, using the appropriate sub-headings (`### Added`, `### Removed`, `### Fixed`), and the links to the appropriate issues/PRs have been included

### CODE REVIEW

File: test/Serialization.purs [hunk 1]
```diff
@@ -83,36 +83,39 @@ suite = do
...
       test
         "PlutusData #6 - Integer 0 (regression to https://github.com/Plutonomicon/cardano-transaction-lib/issues/488 ?)"
         $ do
             let
-              datum = PD.Integer (BigInt.fromInt 0)
-            datum' <- errMaybe "Cannot convertPlutusData" $ convertPlutusData
-              datum
-            let bytes = toBytes (asOneOf datum')
+              datum = convertPlutusData $ PD.Integer (BigInt.fromInt 0)
+            let bytes = toBytes (asOneOf datum)
```
Comment (klntsky) [inline test/Serialization.purs:118]: Use `serializeData`

**Review by klntsky — CHANGES_REQUESTED** [review]
Use `serializeData` everywhere to avoid `asOneOf` call if possible

File: src/Internal/QueryM.purs [hunk 2]
```diff
@@ -818,35 +818,35 @@ applyArgs script args =
...
-          <<< byteArrayToHex
-          <<< Serialization.toBytes
-          <<< asOneOf
-      )
+    ( encodeAeson
+        <<< byteArrayToHex
+        <<< Serialization.toBytes
+        <<< asOneOf
+    )
       <<< Serialization.convertPlutusData
```
Comment (amirmrad) [inline src/Internal/QueryM.purs:850]: Can use `serializeData` https://github.com/Plutonomicon/cardano-transaction-lib/pull/1288#pullrequestreview-1204330512

File: src/Internal/Serialization.purs [hunk 3]
```diff
@@ -871,9 +870,9 @@ hashScriptData cms rs ps = do
   -- function, the resulting hash will be wrong
   case ps of
     [] -> _hashScriptDataNoDatums rs' cms'
-    _ -> _hashScriptData rs' cms' =<< fromJustEff "failed to convert datums"
-      (traverse convertPlutusData ps)
+    _ -> _hashScriptData rs' cms' =<<
+      pure (map convertPlutusData ps)
```
Comment (amirmrad) [inline src/Internal/Serialization.purs:874]: Can be simplified to
```
_hashScriptData rs' cms' $ map convertPlutusData ps
```

File: src/Internal/Serialization/PlutusData.purs [hunk 4]
```diff
@@ -23,38 +23,39 @@ import Ctl.Internal.Types.BigNum (fromBigInt) as BigNum
...
+-- Unsafe fromJust here is correct, because we cover every PlutusData
+-- constructor, and Just will be returned by one of functions
+convertPlutusData x = unsafePartial $ fromJust $ case x of
   T.Constr alt list -> convertConstr alt list
-  T.Map mp -> convertPlutusMap mp
-  T.List lst -> convertPlutusList lst
+  T.Map mp -> Just $ convertPlutusMap mp
+  T.List lst -> Just $ convertPlutusList lst
   T.Integer n -> convertPlutusInteger n
   T.Bytes b -> pure $ _mkPlutusData_bytes b
```
Comment (amirmrad) [inline src/Internal/Serialization/PlutusData.purs:39]: Use either `pure` or `Just` for consistency

File: examples/ContractTestUtils.purs [hunk 5]
```diff
@@ -87,7 +87,7 @@ mkAssertions params@(ContractParams p) = do
     liftedM "Failed to get sender address" $ head <$> getWalletAddresses
   receiverAddress <-
     liftedM "Failed to get receiver address" (getReceiverAddress params)
-  dhash <- liftContractM "Failed to hash datum" $ datumHash $ p.datumToAttach
+  let dhash = datumHash $ p.datumToAttach
```
Comment (jy14898) [inline examples/ContractTestUtils.purs:90]: Not your doing but might as well change it too
```suggestion
  let dhash = datumHash p.datumToAttach
```

File: examples/SatisfiesAnyOf.purs [hunk 6]
```diff
@@ -40,10 +40,8 @@ wrongDatum = Datum $ Integer $ BigInt.fromInt 42
 
 testMustSatisfyAnyOf :: Contract () Unit
 testMustSatisfyAnyOf = do
-  wrongDatumHash <- liftMaybe (error "Cannot get DatumHash") $ Hashing.datumHash
-    wrongDatum
-  correctDatumHash <- liftMaybe (error "Cannot get DatumHash") $
-    Hashing.datumHash unitDatum
+  let wrongDatumHash = Hashing.datumHash wrongDatum
+  let correctDatumHash = Hashing.datumHash unitDatum
   let
```
Comment (jy14898) [inline examples/SatisfiesAnyOf.purs:45]: Merge `let`s (also in other places)

File: src/Internal/Deserialization/PlutusData.purs [hunk 7]
```diff
@@ -46,9 +49,9 @@ convertPlutusData pd =
 convertPlutusConstr :: PlutusData -> Maybe T.PlutusData
 convertPlutusConstr pd = do
   constr <- _PlutusData_constr maybeFfiHelper pd
-  data' <- traverse convertPlutusData
-    $ _unpackPlutusList containerHelper
-    $ _ConstrPlutusData_data constr
+  let
+    data' = convertPlutusData <$>
+      (_unpackPlutusList containerHelper $ _ConstrPlutusData_data constr)
```
Comment (jy14898) [inline src/Internal/Deserialization/PlutusData.purs:54]:
```suggestion
  let
    data' = convertPlutusData <$>
      _unpackPlutusList containerHelper (_ConstrPlutusData_data constr)
```

File: src/Internal/Deserialization/PlutusData.purs [hunk 8]
```diff
@@ -57,16 +60,17 @@ convertPlutusMap pd = do
   entries <- _PlutusData_map maybeFfiHelper pd >>=
     _unpackPlutusMap containerHelper Tuple >>> traverse
       \(k /\ v) -> do
-        k' <- convertPlutusData k
-        v' <- convertPlutusData v
+        let k' = convertPlutusData k
+        let v' = convertPlutusData v
         pure (k' /\ v')
```
Comment (jy14898) [inline src/Internal/Deserialization/PlutusData.purs:65]: Traversing with a pure function, can be rewritten with a map (also let merge)

File: src/Internal/Types/ScriptLookups.purs [hunk 9]
```diff
@@ -422,10 +422,10 @@ validatorM :: forall (a :: Type). Validator -> Maybe (ScriptLookups a)
 validatorM = pure <<< validator
 
 -- | A script lookups value with a datum.
-datum :: forall (a :: Type). Datum -> Maybe (ScriptLookups a)
+datum :: forall (a :: Type). Datum -> ScriptLookups a
 datum dt =
   Hashing.datumHash dt
-    <#> \dh -> over ScriptLookups _ { datums = singleton dh dt } mempty
+    # \dh -> over ScriptLookups _ { datums = singleton dh dt } mempty
```
Comment (amirmrad) [inline src/Internal/Types/ScriptLookups.purs:428]: Can be simplified to
```
over ScriptLookups _ { datums = singleton (Hashing.datumHash dt) dt } mempty
```

File: src/Internal/Types/TypedTxOut.purs [hunk 10]
```diff
@@ -175,23 +175,20 @@ mkTypedTxOut
...
-            -- https://github.com/Plutonomicon/cardano-transaction-lib/issues/691
-            , datum: OutputDatumHash dHash
-            , scriptRef: Nothing
-            }
+    Just <<< mkTypedTxOut' dt $
+      wrap
+        { address
+        , amount
+        -- TODO: populate properly
+        -- https://github.com/Plutonomicon/cardano-transaction-lib/issues/691
```
Comment (amirmrad) [inline src/Internal/Types/TypedTxOut.purs:188]: Not sure if this is needed anymore, the [issue](https://github.com/Plutonomicon/cardano-transaction-lib/issues/691) is closed

File: test/Data.purs [hunk 11]
```diff
@@ -552,8 +552,9 @@ instance (FromData a) => FromData (Tree a) where
   fromData x = genericFromData x
 
 fromBytesFromData :: forall a. FromData a => String -> Maybe a
-fromBytesFromData binary = fromData =<< PDD.convertPlutusData =<< fromBytes
-  (hexToByteArrayUnsafe binary)
+fromBytesFromData binary = fromData =<< (Just <<< PDD.convertPlutusData) =<<
```
Comment (amirmrad) [inline test/Data.purs:555]: Can use `map` or `<$>` instead of `(Just <<< ...)`

File: src/Internal/Deserialization/PlutusData.purs [hunk 12]
```diff
@@ -76,7 +80,8 @@ convertPlutusBytes :: PlutusData -> Maybe T.PlutusData
 convertPlutusBytes pd = T.Bytes <$> _PlutusData_bytes maybeFfiHelper pd
 
 deserializeData :: forall (a :: Type). FromData a => CborBytes -> Maybe a
-deserializeData = (fromData <=< convertPlutusData <=< fromBytes) <<< unwrap
+deserializeData = (fromData <=< (Just <<< convertPlutusData) <=< fromBytes) <<<
+  unwrap
```
Comment (jy14898) [inline src/Internal/Deserialization/PlutusData.purs:84]:
```suggestion
deserializeData = fromData <<< convertPlutusData <=< fromBytes <<< unwrap
```

File: src/Internal/Deserialization/UnspentOutput.purs [hunk 13]
```diff
@@ -98,8 +98,8 @@ convertOutput output = do
     mbDatum = getPlutusData maybeFfiHelper output
   datum <- case mbDatum, mbDataHash of
     Just _, Just _ -> Nothing -- impossible, so it's better to fail
-    Just datumValue, Nothing -> OutputDatum <<< wrap <$> convertPlutusData
-      datumValue
+    Just datumValue, Nothing -> pure <<< OutputDatum <<< wrap $
+      convertPlutusData datumValue
```
Comment (jy14898) [inline src/Internal/Deserialization/UnspentOutput.purs:102]:
```suggestion
    Just datumValue, Nothing -> pure $ OutputDatum $ wrap $
      convertPlutusData datumValue
```

File: src/Internal/Deserialization/WitnessSet.purs [hunk 14]
```diff
@@ -66,8 +66,9 @@ convertWitnessSet :: TransactionWitnessSet -> Maybe T.TransactionWitnessSet
 convertWitnessSet ws = do
   nativeScripts <- for (getNativeScripts maybeFfiHelper ws) convertNativeScripts
   redeemers <- for (getRedeemers maybeFfiHelper ws) convertRedeemers
-  plutusData <- for (getWitnessSetPlutusData maybeFfiHelper ws)
-    convertPlutusList
+  let
+    plutusData = map convertPlutusList
+      (getWitnessSetPlutusData maybeFfiHelper ws)
```
Comment (jy14898) [inline src/Internal/Deserialization/WitnessSet.purs:71]:
```suggestion
  let
    plutusData = convertPlutusList <$> getWitnessSetPlutusData maybeFfiHelper ws
```

**Review by amirmrad — CHANGES_REQUESTED** [review]

File: src/Internal/QueryM.purs [hunk 15]
```diff
@@ -818,35 +818,35 @@ applyArgs script args =
...
+
+            reqBody :: Aeson
+            reqBody = encodeAeson
+              $ Object.fromFoldable
+                  [ "script" /\ scriptToAeson script
+                  , "args" /\ encodeAeson ps
+                  ]
+          liftAff (postAeson url reqBody)
+            <#> map (PlutusScript <<< flip Tuple language) <<<
+              handleAffjaxResponse
```
Comment (jy14898) [inline src/Internal/QueryM.purs:841]:
```suggestion
    Just config -> do
      let
        ps = map plutusDataToAeson args

        language :: Language
        language = snd $ unwrap script

        url :: String
        url = mkHttpUrl config <</>> "apply-args"

        reqBody :: Aeson
        reqBody = encodeAeson
          $ Object.fromFoldable
              [ "script" /\ scriptToAeson script
              , "args" /\ encodeAeson ps
              ]
        liftAff (postAeson url reqBody)
          <#> map (PlutusScript <<< flip Tuple language) <<<
            handleAffjaxResponse
```

File: src/Internal/QueryM.purs (hunk 2 above)
Comment (jy14898) [inline src/Internal/QueryM.purs:850]:
```suggestion
    encodeAeson
      <<< byteArrayToHex
      <<< Serialization.toBytes
      <<< asOneOf
      <<< Serialization.convertPlutusData
```

File: src/Internal/Serialization.purs (hunk 3 above)
Comment (jy14898) [inline src/Internal/Serialization.purs:874]:
```suggestion
    _ -> _hashScriptData rs' cms' (map convertPlutusData ps)
```

File: src/Internal/Serialization.purs [hunk 16]
```diff
@@ -793,8 +793,7 @@ convertTxOutput
         transactionOutputSetDataHash txo
     OutputDatum datumValue -> do
       transactionOutputSetPlutusData txo
-        =<< fromJustEff "convertTxOutput"
-          (convertPlutusData $ unwrap datumValue)
+        =<< pure (convertPlutusData $ unwrap datumValue)
```
Comment (jy14898) [inline src/Internal/Serialization.purs:796]:
```suggestion
        $ convertPlutusData $ unwrap datumValue
```

**Review by jy14898 — CHANGES_REQUESTED** [review]
Looks like Amir has beaten me to most of these changes, I'll make another in a sec, but posting now to avoid more duplication

File: src/Internal/Serialization/PlutusData.purs [hunk 17]
```diff
@@ -23,38 +23,39 @@ import Ctl.Internal.Types.BigNum (fromBigInt) as BigNum
...
+import Data.Maybe (Maybe(Just), fromJust)
 import Data.Tuple (Tuple, fst, snd)
 import Data.Tuple.Nested (type (/\), (/\))
+import Partial.Unsafe (unsafePartial)
 
-convertPlutusData :: T.PlutusData -> Maybe PlutusData
-convertPlutusData = case _ of
+convertPlutusData :: T.PlutusData -> PlutusData
+-- Unsafe fromJust here is correct, because we cover every PlutusData
+-- constructor, and Just will be returned by one of functions
```
Comment (jy14898) [inline src/Internal/Serialization/PlutusData.purs:33]: This isn't true, convertConstr can fail due to BigInt -> BigNum conversion failure

File: src/Internal/Serialization/WitnessSet.purs [hunk 18]
```diff
@@ -129,9 +129,7 @@ convertRedeemer (T.Redeemer { tag, index, "data": data_, exUnits }) = do
   newRedeemer tag' index' data' exUnits'
 
 convertPlutusDataEffect :: PD.PlutusData -> Effect PDS.PlutusData
```
Comment (jy14898) [inline src/Internal/Serialization/WitnessSet.purs:131]: We can just remove this function and use convertPlutusData directly where this function was use. It's exported internally but CTL doesn't use it

File: src/Internal/TxOutput.purs [hunk 19]
```diff
@@ -106,17 +106,17 @@ ogmiosDatumToDatum :: String -> Maybe Datum
 ogmiosDatumToDatum =
   hexToByteArray
     >=> fromBytes
-    >=> Deserialization.convertPlutusData
+    >=> (Deserialization.convertPlutusData >>> Just)
       >>> map Datum
```
Comment (jy14898) [inline src/Internal/TxOutput.purs:110]:
```suggestion
    >=> (Deserialization.convertPlutusData >>> Datum >>> Just)
```

File: src/Internal/Types/ScriptLookups.purs [hunk 20]
```diff
@@ -1459,8 +1459,7 @@ processConstraint mpsMap osMap = do
          OutputDatum
   outputDatum dat = case _ of
     DatumInline -> pure $ OutputDatum dat
-    DatumWitness -> OutputDatumHash <$> liftMaybe (CannotHashDatum dat)
-      (Hashing.datumHash dat)
+    DatumWitness -> pure $ OutputDatumHash $ Hashing.datumHash dat
```
Comment (jy14898) [inline src/Internal/Types/ScriptLookups.purs:1462]: Both cases are pure, can be taken out of the monad

File: src/Internal/Types/TypedTxOut.purs [hunk 21]
```diff
@@ -175,23 +175,20 @@ mkTypedTxOut
   -> Maybe (TypedTxOut validator datum)
 mkTypedTxOut networkId typedVal dt amount =
```
Comment (jy14898) [inline src/Internal/Types/TypedTxOut.purs:176]: Another now pure function, doesn't need to be in Maybe

File: test/Data.purs (hunk 11 above)
Comment (jy14898) [reply test/Data.purs:555]: For cases like this, you can just compose

```suggestion
fromBytesFromData binary = fromData <<< PDD.convertPlutusData =<<
```

**Review by jy14898 — CHANGES_REQUESTED** [review]

File: src/Internal/Serialization/PlutusData.purs (hunk 17 above)
Comment (klntsky) [reply src/Internal/Serialization/PlutusData.purs:33]: Hmmm, no reason to use BigInt then, right? Let's switch @uhbif19

File: src/Internal/Types/TypedTxOut.purs (hunk 10 above)
Comment (uhbif19) [reply src/Internal/Types/TypedTxOut.purs:188]: @klntsky Do we need to just remove the comment or do something? Issue description is blank.

**Review by Bradley-Heather — CHANGES_REQUESTED** [review]

File: src/Internal/Serialization/PlutusData.purs (hunk 17 above)
Comment (uhbif19) [reply src/Internal/Serialization/PlutusData.purs:32]: @klntsky Switched. But there is still `convertPlutusInteger` case. There is conversion from one BigInt to another BigInt which leads to Maybe. But I do not understand what is the difference between BigInts so do not know if it safe to remove Maybe.

Comment (jy14898) [reply src/Internal/Serialization/PlutusData.purs:32]: Looked into it, made a change request

File: src/Internal/Serialization/PlutusData.purs [hunk 22]
```diff
@@ -63,9 +63,9 @@ convertPlutusInteger n =
 convertBigInt :: BigInt.BigInt -> Maybe BigInt
 convertBigInt n = _bigIntFromString maybeFfiHelper (BigInt.toString n)
```
[the remark this answers is not in the document]
Comment (jy14898) [reply src/Internal/Serialization/PlutusData.purs:64]: Both BigInt's are dynamic sized so range errors are not a concern

**Review by jy14898 — COMMENTED** [review]
LGTM

I'm not sure what public functions have been affected by these changes, but clearly at least `datumHash`

Maybe something like this for the changelog?

```
### Changed

- Internal datum conversions are now total, resulting in some datum-related Contract functions dropping the use of `Maybe`, for example `datumHash` ([#1284](https://github.com/Plutonomicon/cardano-transaction-lib/issues/1284)).
```

@klntsky Is there something better we can write, or should we forgo adding this to the changelog?

File: src/Internal/Deserialization/Transaction.purs [hunk 23]
```diff
@@ -262,7 +261,7 @@ convertTxBody txBody = do
     (map <<< map) (M.fromFoldable <<< map (lmap T.RewardAddress))
       -- bignum -> coin
       <<< (traverse <<< traverse <<< traverse)
-        (BigNum.toBigInt' "txbody withdrawals" >>> map Coin)
+        (pure <<< BigNum.toBigInt >>> Coin)
```
Comment (jy14898) [inline src/Internal/Deserialization/Transaction.purs:264]: Lets rewrite this only using `<<<`

File: src/Internal/Deserialization/Transaction.purs [hunk 24]
```diff
@@ -246,9 +246,8 @@ convertTxBody txBody = do
   outputs <-
     _txBodyOutputs containerHelper txBody
       # traverse (convertOutput >>> cslErr "TransactionOutput")
-  fee <-
-    Coin <$> (_txBodyFee txBody # BigNum.toBigInt' "Tx fee")
   let
+    fee = Coin $ (_txBodyFee txBody # BigNum.toBigInt)
```
Comment (jy14898) [inline src/Internal/Deserialization/Transaction.purs:250]:
```suggestion
    fee = Coin $ BigNum.toBigInt $ _txBodyFee txBody
```

File: src/Internal/Deserialization/Transaction.purs [hunk 25]
```diff
@@ -473,31 +472,27 @@ convertProtocolParamUpdate cslPpu = do
   let
     ppu = _unpackProtocolParamUpdate maybeFfiHelper cslPpu
     lbl = (<>) "ProtocolParamUpdate."
-
-  minfeeA <- traverse (map Coin <<< BigNum.toBigInt' (lbl "minfeeA"))
-    ppu.minfeeA
-  minfeeB <- traverse (map Coin <<< BigNum.toBigInt' (lbl "minfeeB"))
-    ppu.minfeeB
+    minfeeA = map (Coin <<< BigNum.toBigInt) ppu.minfeeA
+    minfeeB = map (Coin <<< BigNum.toBigInt) ppu.minfeeB
```
Comment (jy14898) [inline src/Internal/Deserialization/Transaction.purs:476]:
```suggestion
    minfeeA = Coin <<< BigNum.toBigInt <$> ppu.minfeeA
    minfeeB = Coin <<< BigNum.toBigInt <$> ppu.minfeeB
```

File: src/Internal/Deserialization/Transaction.purs [hunk 26]
```diff
@@ -473,31 +472,27 @@ convertProtocolParamUpdate cslPpu = do
...
   maxTxSize <- traverse (cslNumberToUInt (lbl "maxTxSize")) ppu.maxTxSize
   maxBlockHeaderSize <- traverse (cslNumberToUInt (lbl "maxBlockHeaderSize"))
     ppu.maxBlockHeaderSize
-  keyDeposit <- traverse (map Coin <<< BigNum.toBigInt' (lbl "keyDeposit"))
-    ppu.keyDeposit
-  poolDeposit <- traverse (map Coin <<< BigNum.toBigInt' (lbl "poolDeposit"))
-    ppu.poolDeposit
+  let
+    keyDeposit = map (Coin <<< BigNum.toBigInt) ppu.keyDeposit
+    poolDeposit = map (Coin <<< BigNum.toBigInt) ppu.poolDeposit
```
Comment (jy14898) [inline src/Internal/Deserialization/Transaction.purs:484]: Same as the other suggestion, `<$>`

File: src/Internal/Deserialization/Transaction.purs [hunk 27]
```diff
@@ -473,31 +472,27 @@ convertProtocolParamUpdate cslPpu = do
...
   costModels <- addErrTrace (lbl "costModels") $ traverse convertCostModels
     ppu.costModels
-  maxTxExUnits <- traverse (convertExUnits (lbl "maxTxExUnits"))
-    ppu.maxTxExUnits
-  maxBlockExUnits <- traverse (convertExUnits (lbl "maxBlockExUnits"))
-    ppu.maxBlockExUnits
+  let
+    maxTxExUnits = map (convertExUnits (lbl "maxTxExUnits")) ppu.maxTxExUnits
+    maxBlockExUnits = map (convertExUnits (lbl "maxBlockExUnits"))
+      ppu.maxBlockExUnits
```
Comment (jy14898) [inline src/Internal/Deserialization/Transaction.purs:495]: Same as the other suggestion, `<$>`

File: CHANGELOG.md [hunk 28]
```diff
@@ -40,9 +40,12 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 
 - Running plutip servers attaches on SIGINT handlers and therefore node will not exit by default. ([#1231](https://github.com/Plutonomicon/cardano-transaction-lib/pull/1231)).
 - `TestPlanM`, `interpret` and `interpretWithConfig` are now public in `Contract.Test.Mote` and our custom `consoleReporter` in `Contract.Test.Mote.ConsoleReporter`. ([#1261](https://github.com/Plutonomicon/cardano-transaction-lib/pull/1261)).
+- Internal datum conversions are now total, resulting in some datum-related Contract functions dropping the use of `Maybe`, for example `datumHash`, `convertPlutusData` (and related functions). The same with `BigNum.toBigInt`. ([#1284](https://github.com/Plutonomicon/cardano-transaction-lib/issues/1284)).
 
 ### Removed
 
+- `BigNum.toBigIntUnsafe` is dropped.  ([#1284](https://github.com/Plutonomicon/cardano-transaction-lib/issues/1284)).
+
```
Comment (jy14898) [inline CHANGELOG.md:48]: I think we can omit mentions of `BigNum.toBigInt*`, we don't export these functions AFAIK

**Review by jy14898 — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 28 above)
Comment (uhbif19) [reply CHANGELOG.md:48]: They was exported by the module. You mean something else?

File: src/Internal/Deserialization/Transaction.purs (hunk 26 above)
Comment (uhbif19) [reply src/Internal/Deserialization/Transaction.purs:484]: Why <$> is better?

Comment (jy14898) [reply src/Internal/Deserialization/Transaction.purs:484]: It's not significantly better, but you only tend to see map used when the operator would require parenthesis, or when being partially applied

```suggestion
  let
    keyDeposit = Coin <<< BigNum.toBigInt <$> ppu.keyDeposit
    poolDeposit = Coin <<< BigNum.toBigInt <$> ppu.poolDeposit
```

File: CHANGELOG.md (hunk 28 above)
Comment (jy14898) [reply CHANGELOG.md:48]: I mean we don't export it under the `Contract.*` namespace, it's only in internal modules and therefore end users don't need to concern themselves with such changes.

Comment (uhbif19) [reply CHANGELOG.md]: Thanks! Fixed.

File: CHANGELOG.md [hunk 29]
```diff
@@ -45,6 +45,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 
 - Running plutip servers attaches on SIGINT handlers and therefore node will not exit by default. ([#1231](https://github.com/Plutonomicon/cardano-transaction-lib/pull/1231)).
 - `TestPlanM`, `interpret` and `interpretWithConfig` are now public in `Contract.Test.Mote` and our custom `consoleReporter` in `Contract.Test.Mote.ConsoleReporter`. ([#1261](https://github.com/Plutonomicon/cardano-transaction-lib/pull/1261)).
+- Internal datum conversions are now total, resulting in some datum-related Contract functions dropping the use of `Maybe`, for example `datumHash`, `convertPlutusData` (and related functions). The same with `BigNum.toBigInt`. ([#1284](https://github.com/Plutonomicon/cardano-transaction-lib/issues/1284)).
```
Comment (jy14898) [inline CHANGELOG.md:48]:
```suggestion
- Internal datum conversions are now total, resulting in some datum-related Contract functions dropping the use of `Maybe`, for example `datumHash`, `convertPlutusData` and their related functions. ([#1284](https://github.com/Plutonomicon/cardano-transaction-lib/issues/1284)).
```

**Review by jy14898 — CHANGES_REQUESTED** [review]
Last change from me. CI checks are failing though (maybe some will be resolved after merging develop)

**Comment by uhbif19** [comment]
@jy14898 I thought CI was still broken so did not look at them. Thanks.

File: CHANGELOG.md [hunk 30]
```diff
@@ -49,7 +49,8 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 
 - Running plutip servers attaches on SIGINT handlers and therefore node will not exit by default. ([#1231](https://github.com/Plutonomicon/cardano-transaction-lib/pull/1231)).
 - `TestPlanM`, `interpret` and `interpretWithConfig` are now public in `Contract.Test.Mote` and our custom `consoleReporter` in `Contract.Test.Mote.ConsoleReporter`. ([#1261](https://github.com/Plutonomicon/cardano-transaction-lib/pull/1261)).
-- CIP-25 `policy_id` and `asset_name` metadata keys no longer include a `0x` prefix for compatibility with Blockfrost ([#1309](https://github.com/Plutonomicon/cardano-transaction-lib/issues/1309)).
+- Internal datum conversions are now total, resulting in some datum-related Contract functions dropping the use of `Maybe`, for example `datumHash`, `convertPlutusData` and their related functions. ([#1284](https://github.com/Plutonomicon/cardano-transaction-lib/issues/1284)).
+- CIP-25 `policy_id` and `asset_name` metadata keys no longer include a `0x` prefix for compatibility with Blockfrost ([#1309](https://github.com/Plutonomicon/cardano-transaction-lib/issues/1309 "CTL's CIP25 metadata encoding is considered invalid by Blockfrost #1309")).
```
Comment (jy14898) [inline CHANGELOG.md:53]:
```suggestion
- CIP-25 `policy_id` and `asset_name` metadata keys no longer include a `0x` prefix for compatibility with Blockfrost ([#1309](https://github.com/Plutonomicon/cardano-transaction-lib/issues/1309)).
```

**Review by jy14898 — APPROVED** [review]

**Pull request merged**